docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+is authorized to slash or seize the stake.
+
+The authorizer decides up to which amount the given application is eligible to
+slash. This allows managing the risk profile of the staker individually for each
+application. The authorized stake amount can only be increased.
+
+Authorization can not be canceled.
+
+The application being authorized is notified via a callback every single time
+new authorization happens and the application may decide to revert the
+transaction if individual requirements of the application are not met. Such
+requirements could be, for example, not enough stake authorized for the
+particular application or not supported legacy stake (see Limitations section).
+
+All applications first need to be authorized in a registry of T applications
+controlled by the governance.
+
+==== Stake top-up
+
+Top-up increases the amount of tokens locked in a delegation. This increases the
+probability of being chosen for work in the network but is only effective for
+future checks of the stake amount.
+
+Top-ups can be executed for native T stakes and for legacy KEEP/NU stakes.
+
+Native T stakers can only top-up their stakes with a liquid T.
+
+Existing KEEP and NU stakers wanting to execute a top-up have two options. One
+option is to wrap their KEEP/NU to T and then, execute a top-up in T staking
+contract. The second option is to execute a top-up in their legacy staking
+contracts and notify T staking contract about the fact their legacy stake
+increased.
+
+Effectively, it means that existing KEEP stakers can mix their legacy KEEP
+stakes with liquid T stakes. Similarly, existing NU stakers can mix their legacy
+NU stakes with liquid T stakes. This functionality adds some complexity to the
+staking contract but it puts existing KEEP/NU stakers in the same spot as new T
+stakers. Without it, existing stakers would not be able to top-up their stakes
+with T earned from operating in the network, so they would be in a worse spot
+than new T stakers allowed to top-up their stakes using earned T and this way
+increasing their rewards.
+
+Anyone can execute a stake top-up for an operator using a liquid T. Stake top-up
+does not automatically increase authorization levels for applications.
+
+==== Undelegating a stake
+
+The owner or operator may choose to undelegate the stake by submitting an
+undeleagation order. Stake delegation should be marked in the staking contract
+as undelegating. Always the entire stake is getting undelegated and there is a
+fixed, non-governable, two months undelegation period. All applications should
+be designed in such a way that a two-month undelegation period is sufficient to
+retire all groups undelegating operators are involved in. The operator is not
+eligible for any new work during the stake undelegation period but it is still
+required to finish all the pending work. Authorized applications are eligible
+to slash the operator for the entire undelegation period.
```

Comment:
Earlier we discuss that this period will be set per application basis and as result deallocation process also will be per application. After this T tokens can be withdrawable. Do we want to have second layer of undelegation?

Reply:
This is similar to the stake locking mechanism of the KEEP staking contract. We could let applications set their own deallocation periods, locking the stake so that the application has to release it or be stopped with the panic button before the staked tokens can be withdrawn. In this case it would also be possible to do partial withdrawals of stake by tracking the allocated amounts and requiring that the stake left after withdrawal must be at least as much as the highest allocation.

In this case we wouldn't need a separate undelegation period on the T staking contract.

The prospect of the stake being undelegated on KEEP without being properly deallocated on T can be prevented by the T staking contract locking the KEEP stake until all deallocations on T are completed.

Reply:
> The prospect of the stake being undelegated on KEEP without being properly deallocated on T can be prevented by the T staking contract locking the KEEP stake until all deallocations on T are completed.

As far as I understand this is not possible because there's no mechanism for the T staking contract to control stake locks on the KEEP staking contract. On top of that, the latter is not upgradeable. This is essentially the reason why we opened https://github.com/threshold-network/solidity-contracts/issues/14, in agreement with @pdyraga. Can you confirm if this is the case?

Reply:
The T staking contract needs to be authorised by each KEEP operator who wishes to stake on it or else it can't slash them, and this authorisation would also let the T staking contract set and release stake locks on the KEEP staking contract unless there's something I'm missing.

Reply:
Stake can not be locked in Keep staking contract [for more than 200 days](https://github.com/keep-network/keep-core/blob/3a4f9ca6ccf4e2713cea52f401a87be49561c592/solidity/contracts/libraries/staking/Locks.sol#L19). I think we can implement per-application deauthorization and get rid of stake delegation cancellation timeout and global undelegation time (see [here](https://github.com/threshold-network/solidity-contracts/pull/15/files/c1e8a0ba7479ffb23fd6881d4bc08f113a8a2c2c#r704368646) please) but partial withdrawals are making everything quite complex. I would personally avoid them for the first version of this contract.

Reply:
The stake locks can be renewed but it would probably make things a bit too complex.

Partial withdrawals can also be left for later, agreed.

Reply:
Here would be my proposal for deallocations and keeping applications up to date (including solving [this problem](https://github.com/threshold-network/solidity-contracts/pull/15/files/c1e8a0ba7479ffb23fd6881d4bc08f113a8a2c2c#r703723815)):

Each application has an allocation which is the actual amount staked for it, not just a risk ceiling. Each operator also has a locally recorded stake.

The allocation can be increased by the authoriser, up to the operator's stake. This notifies the application of an allocation increase with `increaseAllocation(address operator, uint256 newAmount)`.

The allocation can be decreased by the authoriser, down to zero. This calls `requestAllocationDecrease(address operator, uint256 newAmount)` on the application and the requested decrease is marked as pending on the staking contract. The application may then call `approveAllocationDecrease(address operator)` at its discretion which sets the allocation to the pending amount.

If the operator's stake falls below the allocated amount (including the currently active allocations even if a decrease request is pending) for any application, anyone can report this to the staking contract which then calls `involuntaryAllocationDecrease(address operator, uint256 newAmount)` on the affected application. The application must handle this immediately. If this is reported within a short grace period of the stake being changed, the operator is not penalised further. If this is reported after the grace period is over, whoever submits the report is rewarded with some stake seized from the operator who failed to report it within time.

Optionally, we could also let operators top up their stake within the grace period and not face a penalty if they bring their stake back to the allocation.

We keep a record of all top-ups and slashings along with the times they happened at, per operator, which can be queried to check the grace period. To prevent spam and OOG attacks on involuntary allocation decrease reporting (not sure _why_ someone would do that but it creates an unbounded iteration) there should be a limit on how many times stakes can be topped up within the grace period, so that a top-up is rejected if e.g. the 20 most recent stake changes have all been after `block.timestamp - gracePeriod`.

When an operator is slashed on T, their stake is updated (along with a record of when the stake changed) but allocations aren't; this ensures that any application that has to slash a large number of operators at once can still predict the gas cost of doing so.

I'm not entirely sure about what to do when an operator is slashed on a legacy contract and it's reported to T; we could either leave it without a reward and use a bot, or have a constant reward (because we can't calculate grace periods) which is large enough that it will probably get collected but creates a race to report.

Reply:
The proposed mechanism here would essentially move the async discrepancy reporting functionality from the sortition pool to the staking contract, where we can get more information and thus e.g. use penalties with a suitable grace period.

[Question from here](https://github.com/threshold-network/solidity-contracts/pull/15#discussion_r704400260)

> There is an interesting corner case when authorization happens before someone has a chance to notify about discrepancy but I think we can fix it by allowing to notify per application and slashing multiple times. (?)

In my proposal the authorisation would happen, then after the stake has been updated we would have a discrepancy in the operator's authorisation. The grace period would start from when the stake was updated, and the operator can either top up their stake to eliminate the discrepancy, report it themselves, or wait for someone else to report it and grab the reporting reward.

Reply:
One major weakness with this approach is that if an operator gets slashed to zero, there is no way for the staking contract to pay for the reward for reporting a discrepancy, while a sortition pool would be able to use its own maintenance rewards for the purpose. This can be addressed by including the notifier address in the `involuntaryAllocationDecrease()` call, so the application can pay out its own rewards as well.

Reply:
> Each application has an allocation which is the actual amount staked for it, not just a risk ceiling. Each operator also has a locally recorded stake.

Does it mean that I need to split my stake between several applications? If so, I am worried applications paying smaller rewards can get starved with not enough stakers being available.

Consider the example that random beacon pays 10 T weekly and tBTC v2 pays 100 T weekly for a min stake. Normally, I'd stake for both and earn 110 T but given that I have to choose, I'd go to TBTC v2 with my entire stake .

Reply:
> Does it mean that I need to split my stake between several applications? If so, I am worried applications paying smaller rewards can get starved with not enough stakers being available.

No, the allocations are overlapping so you can allocate the maximum stake to every app

Reply:
> The allocation can be decreased by the authoriser (possibly also operator?), down to zero. This calls `requestAllocationDecrease(address operator, uint256 newAmount)` on the application and the requested decrease is marked as pending on the staking contract. The application may then call `approveAllocationDecrease(address operator)` at its discretion which sets the allocation to the pending amount. A new allocation decrease cannot be requested while an existing decrease is pending.

@eth-r This part needs further clarification. Can you explain further what you mean by `"The application may then call ...."`, who in particular is performing that call?

Reply:
The staking contract would not care about what happens on the application side, all it requires is that `approveAllocationDecrease()` is called by the application.

In practice we could e.g. have the application start a timer when it receives the `requestAllocationDecrease()` call and when the required time has passed the operator can notify the application of this which then makes the application call the staking contract.

so the request would go `operator -> staking -> application` and the approval would go `operator -> application -> staking` where `a -> b` means that `a` calls a function on `b`

Reply:
I think this could be achieved without that second call. When staker calls `requestAllocationDecrease(address operator, uint256 newAmount)`, the staking contract will store the new amount and a future timestamp with the time when tokens will be effectively out of the authorization; additionally it performs a callback to the application in case it should reject the allocation decrease or some other specific logic. Later, if the staker wants to withdraw or make any other change to their stake, we check the pending decreases in the staking contract.

Reply:
Let me try to rephrase it and compare it to the original proposition to make sure we are all on the same page.

The original proposition was: the staked amount is a staking-contract-level variable, `staked`. Each application receives an individual authorization, which is a risk ceiling for the given application, `authorized_app1`, `authorized_app2`, and so forth. In this model, `eligibleStake(appN) = min(staked, authorized_appN)`. It means that we do not need to update the authorized amount for each application if one application slashed the `stake`. When someone authorizes an application, we do validate if `authorized <= stake`, but we do not guarantee this will always be the case in the future. Thus, we have to calculate `eligibleStake` as a function of `staked` and `authorized_appN`.

To keep it simple, the original proposition assumed lowering `authorized_appN` is not possible and there is no way to lower the risk ceiling for applications without undelegating and that the undelegation period is global, hardcoded in the staking contract, and all applications need to be designed in a way obeying the undelegation period. 

We decided that it would be nice - or maybe even a must-have - to have a possibility to decrease the authorization (allocation) from a particular application. Quite a logical implication would be to allow each application to have its own cooldown rules for deauthorization (uncallocation).

I *think* we are not that far in the original proposition from having this feature and we do not have to store the current allocation per application and introduce additional mechanisms to keep them up to date. Trying to modify @eth-r's proposal: 
- `staked` amount is still a staking-contract-level variable
- each application receives an individual authorization (risk ceiling), `authorized_appN`
- `eligibleStake(appN) = min(staked, authorized_appN)`, 
- we probably need a separate `authorized_appN` flavor to reflect the pending authorization decrease for the given app to be able to calculate `activeStake`
- authorizer calls `requestAuthorizationDecrease` on the staking contract, staking contract executes a call to the application so that it can update internal state or even revert the transaction if the given application does not support decreasing the authorization
- after some time, the application calls `finalizeAuthorizationDecrease` that updates `authorized_appN` value on the staking contract; the time before this call is performed by the application, and everything that needs to happen before the process is finalized depends on the particular application.

Reply:
> Let me try to rephrase it and compare it to the original proposition to make sure we are all on the same page.
> 
> The original proposition was: the staked amount is a staking-contract-level variable, `staked`. 

Right, although that's under a model where we allow authorized amounts to go out of sync. We're not necessarily pushing for a different model right now, but it's important to note that if we keep them on sync, we don't need this variable.

> Each application receives an individual authorization, which is a risk ceiling for the given application, `authorized_app1`, `authorized_app2`, and so forth. 

Right, although the fact that's a risk ceiling is from the stake contract point of view.  From the applications point of view, the authorized amount is effectively the "stake" (let's call it "application stake" to not confuse it with the principal stake) and it's used not only as a measure of risk (e.g. for slashing purposes), but also as a basis for rewards, fees, or other potential application logic. 

> In this model, `eligibleStake(appN) = min(staked, authorized_appN)`. It means that we do not need to update the authorized amount for each application if one application slashed the `stake`. When someone authorizes an application, we do validate if `authorized <= stake`, but we do not guarantee this will always be the case in the future. Thus, we have to calculate `eligibleStake` as a function of `staked` and `authorized_appN`.

But if we are operating under the assumption that we allow some synchronization flexibility, there's no need for the application to read the `staked` variable to compute the eligibleStake. Also, we still need to answer the big question of how are we going to ensure some degree of synchronization between app and staking contracts after some time.

> To keep it simple, the original proposition assumed lowering `authorized_appN` is not possible and there is no way to lower the risk ceiling for applications without undelegating and that the undelegation period is global, hardcoded in the staking contract, and all applications need to be designed in a way obeying the undelegation period.
> 
> We decided that it would be nice - or maybe even a must-have - to have a possibility to decrease the authorization (allocation) from a particular application. Quite a logical implication would be to allow each application to have its own cooldown rules for deauthorization (uncallocation).

Yes, the fact that authorizations are non-exclusive directly implies a higher profits and risks margin for the same underlying stake, so it makes sense to allow some risk reduction mechanism.